ClosedFunction.lean

```diff
@@ -0,0 +1,192 @@
+`{p : E × EReal | p.1 ∈ s ∧ f p.1 ≤ p.2}` is closed on `s`. -/
+def ClosedOn (f : E → EReal) (s : Set E) : Prop :=
+  IsClosed {p : E × EReal | p.1 ∈ s ∧ f p.1 ≤ p.2}
+
+/-- A function `f : E → EReal` is said to be closed if its epigraph 
+`{p : E × EReal | f p.1 ≤ p.2}` is closed. -/
+def Closed (f : E → EReal) : Prop := 
+  IsClosed {p : E × EReal | f p.1 ≤ p.2}
+
+theorem ClosedOn.closed_epigraph (hf : ClosedOn f s) : 
+  IsClosed {p : E × EReal | p.1 ∈ s ∧ f p.1 ≤ p.2} := hf
+
+theorem Closed.closed_epigraph (hf : Closed f) :
+  IsClosed {p : E × EReal | f p.1 ≤ p.2} := hf
+
+theorem closed_univ_iff : ClosedOn f univ ↔ Closed f := by 
+  simp [ClosedOn, Closed]
+
+/-!
+### Equivalence
+-/
+
+theorem Closed.isClosed_sublevel {f : E → EReal} (hf : Closed f) :
+    ∀ (r : EReal), IsClosed {x | f x ≤ r} :=
+  fun _ => IsSeqClosed.isClosed fun ⦃_⦄ ⦃_⦄ xns cx =>
+    IsClosed.isSeqClosed hf (fun n => xns n) (Tendsto.prod_mk_nhds cx tendsto_const_nhds)
+
+theorem LowerSemicontinuous.Closed {f : E → EReal}
+    (hf : ∀ x, f x ≤ (liminf f <| 𝓝 x)) : Closed f := by
+  apply IsSeqClosed.isClosed
+  intro f' ⟨x', y'⟩ hxy cxy
+  rw [Prod.tendsto_iff] at cxy
+  let x : ℕ -> E := fun (n : ℕ) => (f' n).1
+  have xley : ∀ (n : ℕ), (f ∘ x) n ≤ (f' n).snd :=
+    fun n => by convert hxy n
+  calc
+    f x' ≤ liminf f (𝓝 x') := hf x'
+    _ ≤ liminf (f ∘ x) atTop := by
+      rw[liminf_eq, liminf_eq]
+      exact sSup_le_sSup (fun _ fa => (eventually_iff_seq_eventually.mp fa) x cxy.1)
+    _ ≤ liminf (fun (n : ℕ) => (f' n).2) atTop := liminf_le_liminf (eventually_of_forall xley)
+    _ = y' := (cxy.2).liminf_eq
+
+theorem lowerSemicontinuous_of_isClosed_sublevel {f : E → EReal}
+    (h : ∀ r : EReal, IsClosed {x | f x ≤ r}) :
+    ∀ x , f x ≤ (liminf f <| 𝓝 x) := by
+  by_contra h; push_neg at h
+  obtain ⟨x, hx⟩ := h
+  obtain ⟨t, ⟨ltt, tlt⟩⟩ := exists_between hx
+  apply not_le_of_gt tlt
+  apply isClosed_iff_frequently.mp (h t) x
+  apply frequently_iff.mpr; intro _ hu
+  have h : ∃ᶠ (y : E) in 𝓝 x, f y ≤ t := by
+    apply frequently_iff.mpr; intro _ hu
+    obtain ⟨x, xu, fx⟩ :=
+      (frequently_iff.mp (frequently_lt_of_liminf_lt (by isBoundedDefault) ltt)) hu
+    exact Exists.intro x ⟨xu, LT.lt.le fx⟩
+  obtain ⟨x, xu, fx⟩ := (frequently_iff.mp h) hu
+  exact Exists.intro x ⟨xu, fx⟩
+
+/-- A function is closed if and only if all its sublevel sets are closed. -/
+theorem closed_iff_IsClosed_sublevel {f : E → EReal} :
+    Closed f ↔ ∀ (r : EReal), IsClosed {x | f x ≤ r} :=
+  ⟨fun h => Closed.isClosed_sublevel h,
+    fun h => LowerSemicontinuous.Closed fun x => lowerSemicontinuous_of_isClosed_sublevel h x⟩
+
+/-- A function is closed if and only if it is lower semi-continuous. -/
+theorem closed_iff_LowerSemicontinuous {f : E → EReal} :
+    Closed f ↔ ∀ x, f x ≤ (liminf f <| 𝓝 x) :=
+  ⟨fun h => lowerSemicontinuous_of_isClosed_sublevel fun _ => Closed.isClosed_sublevel h _,
+    fun a => LowerSemicontinuous.Closed a⟩
```

Comment:
Given these equivalences, I think you don't need to introduce the defs `Closed` and `ClosedOn`; you can simply show that LowerSemicontinuous and [LowerSemicontinuousOn](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Semicontinuous.html#LowerSemicontinuousOn) are equivalent to these two notions and also equivalent to `∀ x, f x ≤ (liminf f <| 𝓝 x)` when you the target of f is a ConditionallyCompleteLattice for liminf to apply. When you need to use the `Closed` and `ClosedOn` conditions, you can just rewrite by these equivalences.